File: workplan_filler.pl
#!/usr/bin/env perl

# Generate dated pages for "My Work Plan" booklet and insert the text into it.
#
# This software is not official. By using it you assume all risks and liabilities.

use strict;
use warnings;
use YAML 'LoadFile';


sub parse_yml_file {
    my $filename = shift @_;

    return LoadFile( $filename );
}


sub generate_pages {
    my $parsed_data = shift @_;

    my $layout = parse_yml_file( 'page_layout.yml' );

    foreach my $idx (0..scalar(@$parsed_data)-1) {
        my $entry               = $parsed_data->[$idx];

        my $unparsed_date       = $entry->{'Date'};
        my ($year,$month,$day);
        if($unparsed_date=~/^(?:Mon|Tue|Wed|Thu|Fri),\s*(\d{4})\_(\d+)\_(\d+)/) {
            ($year,$month,$day) = ($1, $2, $3);
        }
        my $plan_text           = (ref($entry->{'Plan'}) eq 'ARRAY') ? join("\n", @{$entry->{'Plan'}}) : $entry->{'Plan'};
        my $done_text           = (ref($entry->{'Done'}) eq 'ARRAY') ? join("\n", @{$entry->{'Done'}}) : $entry->{'Done'};

        my ($template_file, $date_x_offset, $date_y_offset, $gap_1, $gap_2, $text_x_offset, $plan_y_offset, $done_y_offset)
            = @{$layout->{'coord'}[ $idx % 2 ]}
                {'template_file','date_x_offset','date_y_offset','gap_1','gap_2','text_x_offset','plan_y_offset','done_y_offset'};

        my $output_filename = sprintf("WP_%4d_%02d_%02d.jpg", $year, $month, $day);

        print "Generating $output_filename...\n";

        system( 'convert',
                $template_file,
                -fill       => $layout->{'font_colour'},
                -pointsize  => $layout->{'font_size'},
                -draw => sprintf("text %d,%d '%02d%s%02d%s%4d'", $date_x_offset, $date_y_offset, $day, ' 'x$gap_1, $month, ' 'x$gap_2, $year),
                -draw => sprintf("text %d,%d '%s'", $text_x_offset, $plan_y_offset, $plan_text ),
                -draw => sprintf("text %d,%d '%s'", $text_x_offset, $done_y_offset, $done_text ),
                $output_filename,
        );
    }
}


if(scalar(@ARGV) == 1) {
    generate_pages( parse_yml_file( $ARGV[0] ) );
} else {
    die "Please provide your diary.yml filename as the first argument\n";
}

File: workplan_filler.py
#!/usr/bin/env python3

# Generate dated pages for "My Work Plan" booklet and insert the text into it.
#
# This software is not official. By using it you assume all risks and liabilities.

import sys
import yaml
from datetime import datetime
from subprocess import call

def parse_yml_file(filename):
    with open(filename, 'r') as stream:
        try:
            return(yaml.load(stream))
        except yaml.YAMLError as exc:
            print(exc)
            sys.exit(2)

def join_together(input):
    if(type(input) == list):
        return "\n".join(input)
    else:
        return input

def generate_pages(parsed_data):

    layout  = parse_yml_file('page_layout.yml');

    for idx in range(len(parsed_data)):
        entry           = parsed_data[idx]

        parsed_date     = datetime.strptime(entry['Date'], '%a, %Y_%m_%d')
        year            = parsed_date.year
        month           = parsed_date.month
        day             = parsed_date.day
        output_filename = 'WP_%4d_%02d_%02d.jpg' % (year, month, day)

        plan_text       = join_together( entry['Plan'] )
        done_text       = join_together( entry['Done'] )

        layout_page     = layout['coord'][idx % 2]

        (template_file, date_x_offset, date_y_offset, gap_1, gap_2, text_x_offset, plan_y_offset, done_y_offset) = \
            [ layout_page[k] for k in
                ('template_file','date_x_offset','date_y_offset','gap_1','gap_2','text_x_offset','plan_y_offset','done_y_offset')
            ]

        print('Generating %s ...' % output_filename)

        call([  'convert',      template_file,
                '-fill',        str( layout['font_colour'] ),
                '-pointsize',   str( layout['font_size'] ),
                '-draw',        ("text %d,%d '%02d%s%02d%s%4d'" % (date_x_offset, date_y_offset, day, ' ' * gap_1, month, ' ' * gap_2, year)),
                '-draw',        ("text %d,%d '%s'" % (text_x_offset, plan_y_offset, plan_text )),
                '-draw',        ("text %d,%d '%s'" % (text_x_offset, done_y_offset, done_text )),
                output_filename,
        ]);

if sys.argv[1:] :
    generate_pages( parse_yml_file(sys.argv[1]) )
else:
    print("Please provide your diary.yml filename as the first argument")
    sys.exit(1)

